File: python/566.py
"""Project Euler Problem 566: Cake Cutting.

Given a circular cake with icing on top, if we repeatedly cut slices of cake of lengths
360º/a, 360º/b, 360º/√c in that order, each time flipping over the cut slice, let F(a,b,c)
be the minimum number of flips until all the icing returns to the top. Find
Σ_{9≤a<b<c≤N} F(a,b,c).

For convenience, denote 0=0º and 1=360º and interpolate linearly. We start by considering
the point at the top of the cake at 0. Simulate the cuts, and keep track of the positions
(relative to the knife) of this point as we make cuts, until the point returns to the top
of the cake at 0. While performing this simulation, a small region after this point was
never cut, so this entire region always moves together.

This gives the period for this region, but there may be smaller orders at which the region
may be elsewhere along the cake, but the sequence of whether it is above or below the cake
is the same as the sequence if starting at 0. So we determine all of these orders.

After processing this region, we move on to the next region, and so on, until we've
processed the entire cake. We use the generalized Chinese Remainder Theorem to find the
orders at which all the regions have the same sequence of being above or below the cake.

Finally, as an optimization, when we process a contiguous region of cake, we can also
process any region of cake that the region moved to during the simulation after a full
multiple of 3 flips, because they would all have the same period. A tree map data structure
containing (start, end) intervals is used to efficiently handle these queries.
"""

from __future__ import annotations

from dataclasses import dataclass
from math import gcd, isqrt, sqrt
from typing import List, Set


N = 53

sqrt_c: float = 0.0
d: int = 0


@dataclass(frozen=True)
class Rad:
    """Represents (a+b√c)/d in the current context."""

    a: int
    b: int

    def add(self, other: "Rad") -> "Rad":
        """Add two Rad values."""
        return Rad(self.a + other.a, self.b + other.b)

    def subtract(self, other: "Rad") -> "Rad":
        """Subtract two Rad values."""
        return Rad(self.a - other.a, self.b - other.b)

    def min(self, other: "Rad") -> "Rad":
        """Minimum of two Rad values."""
        return self if self.compare_to(other) <= 0 else other

    def max(self, other: "Rad") -> "Rad":
        """Maximum of two Rad values."""
        return self if self.compare_to(other) >= 0 else other

    def __lt__(self, other: "Rad") -> bool:
        """Less than comparison."""
        return (self.a + self.b * sqrt_c) < (other.a + other.b * sqrt_c)

    def __le__(self, other: "Rad") -> bool:
        """Less than or equal comparison."""
        return (self.a + self.b * sqrt_c) <= (other.a + other.b * sqrt_c)

    def compare_to(self, other: "Rad") -> int:
        """Compare two Rad values. Returns -1, 0, or 1."""
        val1 = self.a + self.b * sqrt_c
        val2 = other.a + other.b * sqrt_c
        if val1 < val2:
            return -1
        if val1 > val2:
            return 1
        return 0

    @staticmethod
    def ONE() -> "Rad":
        """Return Rad representing 1."""
        return Rad(d, 0)


@dataclass(frozen=True)
class Pos:
    """Position on cake."""

    angle: Rad
    flipped: bool


def is_sq(n: int) -> bool:
    """Check if n is a perfect square."""
    root = isqrt(n)
    return root * root == n


def lcm(a: int, b: int) -> int:
    """Least common multiple."""
    return a * b // gcd(a, b)


def lin_comb(a: int, b: int) -> Rad:
    """Find solution to ax + by = gcd(a,b) using extended Euclidean algorithm."""
    if b == 0:
        return Rad(1, 0)
    x1 = lin_comb(b, a % b)
    q = a // b
    return Rad(x1.b, x1.a - q * x1.b)


def mod(a: int, m: int) -> int:
    """Modular reduction."""
    return ((a % m) + m) % m


def pow_int(base: int, exp: int) -> int:
    """Integer power."""
    return base**exp


def general_crt(a1: int, m1: int, a2: int, m2: int) -> int:
    """Generalized Chinese Remainder Theorem.

    Returns x such that x≡a1 (mod m1) and x≡a2 (mod m2), where m1 and m2 need not be co-prime.
    Returns -1 if no solution exists.
    """
    g = gcd(m1, m2)
    if a1 % g != a2 % g:
        return -1
    lin = lin_comb(m1, m2)
    return mod(a1 - m1 * lin.a * (a1 - a2) // g, m1 * m2 // g)


def identical_shifts(vals: List[int], k: int) -> Set[int]:
    """Returns all i such that the values at i, i+k, i+2k, ..., i-k (wrapping around) equal the
    values at 0, k, 2k, ...
    """
    H = 3
    pow_h = pow_int(H, len(vals) // k)
    target_hash = 0
    for i in range(0, len(vals), k):
        target_hash = target_hash * H + vals[i]

    shifts: Set[int] = set()
    for initial_shift in range(k):
        hash_val = 0
        for i in range(initial_shift, len(vals), k):
            hash_val = hash_val * H + vals[i]
        for i in range(initial_shift, len(vals), k):
            if hash_val == target_hash:
                shifts.add(i)
            hash_val = hash_val * H + (1 - pow_h) * vals[i]
    return shifts


class ProcessedRegions:
    """Contains an entry (start, end) for each interval. No two entries intersect."""

    def __init__(self) -> None:
        """Initialize processed regions."""
        self.intervals: List[tuple[Rad, Rad]] = []
        self.point = Rad(0, 0)

    def advance_to_next_unprocessed(self) -> None:
        """Advance to next unprocessed point."""
        # Find interval containing point
        for start, end in self.intervals:
            if start.compare_to(self.point) <= 0 and self.point.compare_to(end) < 0:
                self.point = end
                return
        # If not found, point is at end of last interval or at 0
        if self.intervals:
            self.point = self.intervals[-1][1]
        else:
            self.point = Rad(0, 0)

    def process_interval(self, start: Rad, end: Rad) -> None:
        """Process an interval, merging with existing intervals."""
        # Check if start is within an existing interval
        merged_start = start
        merged_end = end
        new_intervals: List[tuple[Rad, Rad]] = []
        for s, e in self.intervals:
            # Check for overlap: intervals overlap if not (e < start or end < s)
            if not (e.compare_to(start) < 0 or end.compare_to(s) < 0):
                # Overlap - merge
                merged_start = merged_start.min(s)
                merged_end = merged_end.max(e)
            else:
                # No overlap - keep interval
                new_intervals.append((s, e))
        new_intervals.append((merged_start, merged_end))
        # Sort by start value
        new_intervals.sort(key=lambda x: (x[0].a, x[0].b))
        self.intervals = new_intervals

    def done(self) -> bool:
        """Check if all regions processed."""
        return self.point.compare_to(Rad.ONE()) >= 0


def order(flip_sizes: List[Rad]) -> int:
    """Compute order for given flip sizes."""
    orders: Set[int] = {0}
    period = 1
    regions = ProcessedRegions()
    while not regions.done():
        regions.advance_to_next_unprocessed()
        start_pos = Pos(regions.point, False)
        positions: List[Pos] = []
        uncut_region = Rad.ONE()
        pos = start_pos
        first = True
        while first or not (
            pos.angle.compare_to(start_pos.angle) == 0 and pos.flipped == start_pos.flipped
        ):
            first = False
            for size in flip_sizes:
                positions.append(pos)
                remaining = Rad.ONE().subtract(pos.angle)
                if pos.angle.compare_to(size) < (1 if pos.flipped else 0):
                    uncut_region = uncut_region.min(
                        pos.angle if pos.flipped else size.subtract(pos.angle)
                    )
                    pos = Pos(remaining, not pos.flipped)
                else:
                    uncut_region = uncut_region.min(
                        pos.angle.subtract(size) if pos.flipped else remaining
                    )
                    pos = Pos(pos.angle.subtract(size), pos.flipped)

        curr_region_orders = identical_shifts(
            [1 if p.flipped else 0 for p in positions], len(flip_sizes)
        )
        curr_period = len(positions)
        new_orders: Set[int] = set()
        for order_val in orders:
            for curr_order in curr_region_orders:
                new_order = general_crt(order_val, period, curr_order, curr_period)
                if new_order != -1:
                    new_orders.add(new_order)
        period = lcm(period, curr_period)
        orders = new_orders

        for i in range(0, len(positions), len(flip_sizes)):
            pos = positions[i]
            if not pos.flipped:
                regions.process_interval(pos.angle, pos.angle.add(uncut_region))

    orders.discard(0)
    return min(orders) if orders else period


def F(a: int, b: int, c: int) -> int:
    """Compute F(a, b, c)."""
    global sqrt_c, d
    if is_sq(c):
        sqrt_c = 1.0
        d = a * b * isqrt(c)
        return order([Rad(d // a, 0), Rad(d // b, 0), Rad(d // isqrt(c), 0)])
    else:
        sqrt_c = sqrt(c)
        d = a * b * c
        return order([Rad(d // a, 0), Rad(d // b, 0), Rad(0, d // c)])


def solve() -> int:
    """Solve Problem 566."""
    ans = 0
    for a in range(9, N):
        for b in range(a + 1, N):
            for c in range(b + 1, N + 1):
                ans += F(a, b, c)
    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/566.c
/*
 * Project Euler Problem 566: Cake Cutting
 *
 * Given a circular cake with icing on top, if we repeatedly cut slices of
 * lengths 360/a, 360/b, 360/sqrt(c), each time flipping, find F(a,b,c) =
 * minimum flips until all icing returns to top.
 * Compute sum F(a,b,c) for 9 <= a < b < c <= 53.
 *
 * This is a direct C translation of the Python solution using Rad (a+b*sqrt_c)/d
 * representation, generalized CRT, and interval-based region tracking.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MAX_POSITIONS 200000
#define MAX_INTERVALS 50000
#define MAX_ORDERS 100000

static double glob_sqrt_c;
static long long glob_d;

/* Rad represents (a + b*sqrt_c)/d */
typedef struct { long long a; long long b; } Rad;

static Rad rad_add(Rad x, Rad y) { return (Rad){x.a + y.a, x.b + y.b}; }
static Rad rad_sub(Rad x, Rad y) { return (Rad){x.a - y.a, x.b - y.b}; }

static double rad_val(Rad x) { return x.a + x.b * glob_sqrt_c; }

static int rad_cmp(Rad x, Rad y) {
    double vx = rad_val(x), vy = rad_val(y);
    if (vx < vy) return -1;
    if (vx > vy) return 1;
    return 0;
}

static Rad rad_min(Rad x, Rad y) { return rad_cmp(x, y) <= 0 ? x : y; }
static Rad rad_max(Rad x, Rad y) { return rad_cmp(x, y) >= 0 ? x : y; }
static Rad rad_one(void) { return (Rad){glob_d, 0}; }
static Rad rad_zero(void) { return (Rad){0, 0}; }

typedef struct { Rad angle; int flipped; } Pos;

static int is_sq(int n) {
    int r = (int)sqrt((double)n);
    if (r * r == n) return 1;
    if ((r+1)*(r+1) == n) return 1;
    return 0;
}

static int isqrt_i(int n) {
    int r = (int)sqrt((double)n);
    while (r * r > n) r--;
    while ((r+1)*(r+1) <= n) r++;
    return r;
}

static long long gcd_ll(long long a, long long b) {
    if (a < 0) a = -a;
    if (b < 0) b = -b;
    while (b) { long long t = b; b = a % b; a = t; }
    return a;
}

static long long lcm_ll(long long a, long long b) {
    return a / gcd_ll(a, b) * b;
}

static long long mod_ll(long long a, long long m) {
    return ((a % m) + m) % m;
}

/* Extended GCD returning (x, y) such that a*x + b*y = gcd(a,b) */
static void ext_gcd(long long a, long long b, long long *x, long long *y) {
    if (b == 0) { *x = 1; *y = 0; return; }
    long long x1, y1;
    ext_gcd(b, a % b, &x1, &y1);
    *x = y1;
    *y = x1 - (a / b) * y1;
}

/* General CRT: find x such that x = a1 (mod m1) and x = a2 (mod m2) */
static long long general_crt(long long a1, long long m1, long long a2, long long m2) {
    long long g = gcd_ll(m1, m2);
    if ((a1 - a2) % g != 0) return -1;
    long long lx, ly;
    ext_gcd(m1, m2, &lx, &ly);
    long long lcm = m1 / g * m2;
    return mod_ll(a1 - m1 * ((__int128)lx * ((a1 - a2) / g) % lcm), lcm);
}

/* Intervals for processed regions */
typedef struct { Rad start; Rad end; } Interval;

static Interval intervals[MAX_INTERVALS];
static int n_intervals;
static Rad current_point;

static void intervals_init(void) {
    n_intervals = 0;
    current_point = rad_zero();
}

static void advance_to_next_unprocessed(void) {
    for (int i = 0; i < n_intervals; i++) {
        if (rad_cmp(intervals[i].start, current_point) <= 0 &&
            rad_cmp(current_point, intervals[i].end) < 0) {
            current_point = intervals[i].end;
            return;
        }
    }
}

static void process_interval(Rad start, Rad end) {
    Rad ms = start, me = end;
    Interval new_ints[MAX_INTERVALS];
    int nn = 0;
    for (int i = 0; i < n_intervals; i++) {
        /* Check overlap */
        if (rad_cmp(intervals[i].end, start) < 0 || rad_cmp(end, intervals[i].start) < 0) {
            new_ints[nn++] = intervals[i];
        } else {
            ms = rad_min(ms, intervals[i].start);
            me = rad_max(me, intervals[i].end);
        }
    }
    new_ints[nn++] = (Interval){ms, me};
    /* Sort by start */
    for (int i = 0; i < nn - 1; i++)
        for (int j = i + 1; j < nn; j++)
            if (rad_cmp(new_ints[i].start, new_ints[j].start) > 0) {
                Interval tmp = new_ints[i]; new_ints[i] = new_ints[j]; new_ints[j] = tmp;
            }
    memcpy(intervals, new_ints, nn * sizeof(Interval));
    n_intervals = nn;
}

static int intervals_done(void) {
    return rad_cmp(current_point, rad_one()) >= 0;
}

/* identical_shifts: find all i such that vals shifted by i (step k) matches vals at 0,k,2k... */
static int shift_results[MAX_POSITIONS];
static int n_shift_results;

static void identical_shifts(int *vals, int len, int k) {
    n_shift_results = 0;
    int H = 3;
    int sublen = len / k;
    long long pow_h = 1;
    for (int i = 0; i < sublen; i++) pow_h *= H;

    long long target_hash = 0;
    for (int i = 0; i < len; i += k)
        target_hash = target_hash * H + vals[i];

    for (int initial_shift = 0; initial_shift < k; initial_shift++) {
        long long hash_val = 0;
        for (int i = initial_shift; i < len; i += k)
            hash_val = hash_val * H + vals[i];
        for (int i = initial_shift; i < len; i += k) {
            if (hash_val == target_hash)
                shift_results[n_shift_results++] = i;
            hash_val = hash_val * H + (1 - pow_h) * vals[i];
        }
    }
}

static Pos positions[MAX_POSITIONS];
static int n_positions;

static int order_compute(Rad *flip_sizes, int n_flips) {
    /* Orders stored as set of longs */
    static long long orders[MAX_ORDERS];
    static long long new_orders[MAX_ORDERS];
    int n_orders = 1;
    orders[0] = 0;
    long long period = 1;

    intervals_init();

    while (!intervals_done()) {
        advance_to_next_unprocessed();
        if (intervals_done()) break;

        Pos start_pos = {current_point, 0};
        n_positions = 0;
        Rad uncut_region = rad_one();
        Pos pos = start_pos;
        int first = 1;

        while (first || !(rad_cmp(pos.angle, start_pos.angle) == 0 && pos.flipped == start_pos.flipped)) {
            first = 0;
            for (int fi = 0; fi < n_flips; fi++) {
                if (n_positions >= MAX_POSITIONS) break;
                positions[n_positions++] = pos;
                Rad remaining = rad_sub(rad_one(), pos.angle);
                Rad size = flip_sizes[fi];
                int cmp = rad_cmp(pos.angle, size);
                if (cmp < (pos.flipped ? 1 : 0)) {
                    uncut_region = rad_min(uncut_region,
                        pos.flipped ? pos.angle : rad_sub(size, pos.angle));
                    pos = (Pos){remaining, !pos.flipped};
                } else {
                    uncut_region = rad_min(uncut_region,
                        pos.flipped ? rad_sub(pos.angle, size) : remaining);
                    pos = (Pos){rad_sub(pos.angle, size), pos.flipped};
                }
            }
        }

        /* Build flipped array */
        int *flip_arr = (int *)malloc(n_positions * sizeof(int));
        for (int i = 0; i < n_positions; i++)
            flip_arr[i] = positions[i].flipped ? 1 : 0;

        identical_shifts(flip_arr, n_positions, n_flips);

        int curr_period = n_positions;
        int n_new = 0;
        for (int oi = 0; oi < n_orders; oi++) {
            for (int si = 0; si < n_shift_results; si++) {
                long long nv = general_crt(orders[oi], period, shift_results[si], curr_period);
                if (nv != -1) {
                    /* Check if already in new_orders */
                    int found = 0;
                    for (int k = 0; k < n_new; k++)
                        if (new_orders[k] == nv) { found = 1; break; }
                    if (!found && n_new < MAX_ORDERS)
                        new_orders[n_new++] = nv;
                }
            }
        }
        period = lcm_ll(period, curr_period);
        memcpy(orders, new_orders, n_new * sizeof(long long));
        n_orders = n_new;

        /* Process regions */
        for (int i = 0; i < n_positions; i += n_flips) {
            if (!positions[i].flipped) {
                process_interval(positions[i].angle, rad_add(positions[i].angle, uncut_region));
            }
        }

        free(flip_arr);
    }

    /* Find minimum non-zero order */
    long long min_order = period;
    for (int i = 0; i < n_orders; i++) {
        if (orders[i] > 0 && orders[i] < min_order)
            min_order = orders[i];
    }
    return (int)min_order;
}

static int F(int a, int b, int c) {
    Rad flip_sizes[3];
    if (is_sq(c)) {
        int sc = isqrt_i(c);
        glob_sqrt_c = 1.0;
        glob_d = (long long)a * b * sc;
        flip_sizes[0] = (Rad){glob_d / a, 0};
        flip_sizes[1] = (Rad){glob_d / b, 0};
        flip_sizes[2] = (Rad){glob_d / sc, 0};
    } else {
        glob_sqrt_c = sqrt((double)c);
        glob_d = (long long)a * b * c;
        flip_sizes[0] = (Rad){glob_d / a, 0};
        flip_sizes[1] = (Rad){glob_d / b, 0};
        flip_sizes[2] = (Rad){0, glob_d / c};
    }
    return order_compute(flip_sizes, 3);
}

int main(void) {
    int NN = 53;
    long long ans = 0;
    for (int a = 9; a < NN; a++) {
        for (int b = a + 1; b < NN; b++) {
            for (int c = b + 1; c <= NN; c++) {
                ans += F(a, b, c);
            }
        }
    }
    printf("%lld\n", ans);
    return 0;
}
